README-TESTING.md

```diff
@@ -0,0 +1,32 @@
+# Testing the libraries
+
+The code in these libraries is tested both with deductive verification, using dafny,
+and with traditional runtime testing of compilations by dafny. 
+The latter is important for those cases (e.g. FileIO) where the implementations
+are custom external code for each target programming language.
+Runtime tests are also a check that dafny compiled the (verified) dafny code correctly.
+All files are also checked that they conform to standard dafny formatting,
+using `dafny format`.
+
+The tests are run using `lit`. The configuration of the tests is controlled by
+(a) the lit.site.cfg file and (b) the yml files in .github/workflows.
+
+The tests are run on each push to a pull_request and are required to pass to enable merging.
+They are also run on a nightly schedule.
+
+To run the files manually: run `lit .` in this directory (`libraries`, at the top
+of the clone of the repo). You can also manually run individual files or directories,
+but `lit` must be invoked in the `libraries` directory. You can also trigger the workflow
+using `workflow_dispatch`. To do a manual check of formatting, use `dafny format --check <file or folder>`.
+
+The tests themselves are specially formatted lines (those beginning with `// RUN:`) in
+each .dfy file. The tests use dafny's new command-line syntaxi, using the lit
+macros `%verify` and `%run` and the like.
+
+The versions of dafny being run are set in the `tests.yml` file. The versions tested
+ must be recent enough to support the new CLI syntax, so 3.11ff. Also, if 
```

Comment:
Truncated?

Reply:
Odd. Its OK in the Files view. I'll check the rendering again after pushing these changes.